File: 76.minimum-window-substring.java
/*
 * @lc app=leetcode id=76 lang=java
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
class Solution {
    public String minWindow(String s, String t) {
        char[] sChar = s.toCharArray(), tChar = t.toCharArray();
        int[] need = new int[256];
        for (int i = 0; i < tChar.length; i++) {
            need[tChar[i]]++;
        }
        int needCharCNT = tChar.length;
        int left = 0, right = 0;
        int ansL = -1, ansR = -1, len = Integer.MAX_VALUE;
        while (right < sChar.length) {

            need[sChar[right]]--;
            if (need[sChar[right]] >= 0) {
                needCharCNT--;
            }

            // try to make this substring shorter in this loop
            while (needCharCNT == 0) {
                if (right - left < len) {
                    // save the current shortest substring
                    ansL = left;
                    ansR = right;
                    len = ansR - ansL;
                }

                // restore before left+1
                need[sChar[left]]++;
                if (need[sChar[left]] > 0) {
                    needCharCNT++;
                }
                left++;
            }

            right++;
        }
        return ansL == -1 ? "" : s.substring(ansL, ansR + 1);
    }
}
// @lc code=end

File: 76.minimum-window-substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        n, m = len(s), len(t)
        if n < m:
            return ""

        cnt=Counter(t)
        curr_cnt=defaultdict(int)

        filtered_s=[]
        for i in range(n):
            if s[i] in cnt:
                filtered_s.append((i, s[i]))

        left, right = 0, 0
        formed=0
        res=float("inf"), 0,0
        
        while right < len(filtered_s):
            char = filtered_s[right][1]
            curr_cnt[char]+=1

            if curr_cnt[char]==cnt[char]:
                formed+=1

            while left <= right and formed == len(cnt):
                char = filtered_s[left][1]

                start = filtered_s[left][0]
                end = filtered_s[right][0]
                if end-start+1 < res[0]:
                    res=(end-start+1,start,end)
                
                curr_cnt[char]-=1
                if curr_cnt[char]<cnt[char]:
                    formed-=1
                
                left+=1
            
            right+=1

        if res[0] == float("inf"):
            return ""
        else:
            return s[res[1]:res[2]+1]
# @lc code=end

